### ISSUE TITLE
Windows containers: couchdb fails when data directory mounted as a volume

### ISSUE THREAD

**Issue by drnybble** [docker/docker#31524]
**Description**

**Steps to reproduce the issue:**
Grab this attachment and unzip it.

[couchdb_issue.zip](https://github.com/docker/docker/files/817306/couchdb_issue.zip)

Run the build.ps1 script to build the couchdb image.
Run the good.ps1 script -- couchdb starts up successfully when the C:\couchdb\var\lib\couchdb directory is not mounted as a volume.
Run the bad.ps1 script -- couchdb quits on start-up.

**Describe the results you received:**
couchdb does not start up successfully when its data directory is mounted as a volume.

**Describe the results you expected:**
couchdb should start up whether its data directory is mounted as a volume or not.

**Additional information you deem important (e.g. issue happens only occasionally):**
Interestingly it does sometimes work. If I run the bad.ps1 script many times eventually CouchDB was able to create one file (.delete) in the host folder. After many more times it eventually created the other two files and started up.
By observing the ".delete" path in Process Monitor I can see that the couchdb Erlang code makes a call to CreateFileW. Eventually the FLTMGR.SYS driver makes a CreateFile call and gets a REPARSE response. At this point CouchDB quits.
If I try long enough eventually CouchDB is able to create the ".delete" file. In this case I see a subsequent call by the FLTMGR.SYS driver where it gets an "IS_DIRECTORY" response from an additional CreateFile call. So it seems like sometimes FLTMGR.SYS does not issue a call following a REPARSE result.

**Output of `docker version`:**

```
Client:
 Version:      17.04.0-dev
 API version:  1.27
 Go version:   go1.7.5
 Git commit:   0b9fddb
 Built:        Thu Mar  2 17:40:58 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.04.0-dev
 API version:  1.27 (minimum version 1.24)
 Go version:   go1.7.5
 Git commit:   0b9fddb
 Built:        Thu Mar  2 17:40:58 2017
 OS/Arch:      windows/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 304
Server Version: 17.04.0-dev
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: l2bridge l2tunnel nat null overlay transparent
Swarm: inactive
Default Isolation: process
Kernel Version: 10.0 14393 (14393.693.amd64fre.rs1_release.161220-1747)
Operating System: Windows Server 2016 Standard
OSType: windows
Architecture: x86_64
CPUs: 4
Total Memory: 16GiB
Name: WIN-U9UU27TK2FQ
ID: HOYH:CLDY:C2SF:SJJN:VEN6:H2CS:FECI:QKIM:WA3K:3PJX:2QPE:F443
Docker Root Dir: C:\ProgramData\docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 pa-docker:5000
 pa-docker.ottawa.ibm.com:5000
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Running on fully up-to-date Server 2016 inside a VirtualBox VM.

**Comment by thaJeztah** [comment]
ping @jstarks @jhowardmsft @johnstep PTAL

**Comment by drnybble** [comment]
I did try launching couchdb through PowerShell and also put a **sleep 5** before starting. Didn't help.

**Comment by johnstep** [comment]
I have not had a chance, but will try to reproduce this today. Thanks.

**Comment by johnstep** [comment]
There are multiple issues:

1. Erlang/OTP R16B02 (`efile_readlink` in beam.dll, called from erl.exe) attempts to resolve the reparse point (symbolic link to the volume) to its final name via [GetFinalPathNameByHandleW ](https://msdn.microsoft.com/en-us/library/windows/desktop/aa364962(v=vs.85).aspx), which fails with **ERROR_FILE_NOT_FOUND**. `efile_readlink` does not check for error, and corrupts memory.
2. **GetFinalPathNameByHandleW** generally fails to return the final name for volume directories mapped into containers, returning **ERROR_FILE_NOT_FOUND**. Based on debugging, it looks like it resolves to the target path on the **host**, which is invalid in the container.
3. If **GetFinalPathNameByHandleW** succeeded, it would likely return a path of the form `\\?\ContainerMappedDirectories\<GUID>`, a valid, but unconventional Win32 file name. Many cmd.exe commands fail with those file names. Erlang immediately removes the `\\?\` prefix, and assumes the rest is valid Win32 file name. For container mapped directories (volumes), that is not the case, so removing the prefix will result in an invalid file name, e.g. `ContainerMappedDirectories\<GUID>`.

Given these issues, even if Microsoft does update **GetFinalPathNameByHandleW** to succeed for container mapped directories (for example, by returning the `\\?\ContainerMappedDirectories\<GUID>` path), callers may fail parsing it. The best option is for a caller to not modify the name at all.

@drnybble Based on code inspection, Erlang/OTP 19.2 checks for failure from **GetFinalPathNameByHandleW**, allowing for a graceful failure. CouchDB 2.0.0 for Windows is based on Erlang/OTP 17.5. I have not reviewed its source, or tried it, but is 2.0.0 and option for your scenario?

**Comment by drnybble** [comment]
John, thanks for this superb analysis. I will work on Dockerizing couchdb 2.0.0 and see if it works. I guess it was working on 1.6 sometimes by fluke perhaps due to memory corruption as you said.

**Comment by thaJeztah** [comment]
It looks like there's nothing actionable at this moment, so I'll close this issue for housekeeping, but let us know how you progressed 👍

**Comment by johnstep** [comment]
@donkirkby Sounds good. Yes, I noticed if I prevented the corruption, by changing `ecx` to 0 so it did not copy any memory, then CouchDB start successfully. Also, I believe that once the directory structure exists, this code path is not encountered.

I took a quick look at CouchDB 2.0.0.1. It seems quite different, and did not work on my first attempt, though it failed in a completely different way. Looking at the Erlang/OTP 17.5 source code, it _does_ check the **GetFinalPathNameByHandleW** return value, so it should, at least, avoid that problem.
